### DISCUSSION TITLE
Schedule Migration

### DISCUSSION THREAD

**Discussion by trufrgs** [flamingock/mongock#44] [Ideas]
Hello Guys, first of all thanks for the attention. I've looked for any template for opening discussions/suggestions and didn't find one, please let me know if I'm missing something.

We are having internal discussions about our DB migrations, which today we execute through an internal tool, in a way like your CLI. We took a look at mongock and found it very interesting, we made some POCs and it works smoothly for what it is designed to. However in our discussions, we understand that the DB changes do not always have to be synchronous with the server deployment. Shortly explaining the cases we are thinking of are: 

- Data transformation (fixing some application data, changing some flags, etc). We understand that this may not be the main use case intended for this tool, but it would be useful, because we have a control of what already has been executed, we have a organized log history, we guarantee unicity, we can use rollback in case of failing and everything is "managed" by the tool.
- "Schema" migration (removing or changing some field or collection that is used by the current running server): as mentioned in your doc, there are some tricky situations where it is necessary to make a 2-step approach, we understand that when we face that we don't necessarily need to run the migration before running the server, since we have to make a multi-phase strategy in order to guarantee integrity.

That said, we think in most cases the DB migration could be decoupled from the server deployment, and should not affect this. We understand that for a lot of "simpler" situations this is satisfactory and helps to guarantee integrity, mainly in "1 node" running situations (what happens when we migrate db and not all nodes are up with updated code yet?). 

In order to achieve this level of asynchrony I've seen a [relevant discussion here](https://github.com/mongock/mongock/discussions/402), but would be very interesting if we have intentional features on the tool to help manage this Scheduled Migrations and control what have to be run before application startup and what can be run at any time wanted.

**Comment by dieppa** [comment]
Hello @trufrgs 

as you well pointed, the main use of Mongock, or at least the original one, is to ship the migration together with the code. This is not for simple cases, it's actually a common use among microservices and distribute environments and obviously for multi-node service.

Said that, we also think there are also other valid and useful approaches that require decoupling the migration from the application code 

For now we are providing the CLI, which currently requires the application jar. However we are working to improve this, so you can run the CLI directly with your changeUnits, without the application involvement. Or even without anything at all, no even changeUnits, if what you only want is to see the state of the database, for example.

We think that's a big move in the migration autonomy area, although we also have in mind further steps.

However, I have couple of questions for you:

- [This thread](https://github.com/mongock/mongock/discussions/402) you highlighted mentions a way to run the migration asynchronously, but still within the app? It provides some level of decoupling, but is far form what I mentioned. I don't really understand what benefit you would get from this. Can you elaborate it, please?   

- In that same paragraph you mention  `Scheduled Migrations`, but I am not clear what you mean by that. We also had this in the roadmap(low priority tbh), but our idea is about providing `cron` capabilities to migrations. Can you explain what you mean by that? Basically this part: `Scheduled Migrations and control what have to be run before application startup and what can be run at any time wanted.`

**Reply by trufrgs** [reply]
@dieppa thanks for the quick and complete answer, also sorry for the delay :)

I took a deeper look at the CLI tool, and reading what you wrote about improvements you are working on, and I agree with you... this may be exaclty what we were looking for in terms of "declouped db migrations/updates".

Answering the questions:

1. I agree with you and I don't think thats the ideal way towards a migration autonomy. Just mentioned that as a palliative in order to try to make my point clearer.
2. By scheduled migrations I meant exactly someting like a cron capability 😄 . So about the sentence you mentioned, when we have cron capability, I think those "scheduled" change units (in another words, those that are scheduled under a cron job) won't lock the application startup if we deploy them, right? However there will also exist change units that we want to guarantee that are ran before the aplication startup (as the current behavior of the tool)... and that's great. I could set some change units that I want to be ran before app startup and avoid startup if smtg fails, and some others that don't need to lock the app startup and can be scheduled with a cron (if it fails we could just raise an event/alert, but the app keeps running).

I've tried my best to make it clear, but if it is not yet, please let me know. 

In short, I understood that you're evolving the tool into a good way, offering capabilities to run some migrations in a declouped way (with some cron job like feature) and keeping the current (very useful) behavior option: coupled with the application startup.
